fsmonitor.c

```diff
@@ -173,27 +189,55 @@ void refresh_fsmonitor(struct index_state *istate)
 	 * should be inclusive to ensure we don't miss potential changes.
 	 */
 	last_update = getnanotime();
-	strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
+	if (hook_version == HOOK_INTERFACE_VERSION1)
+		strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
 
 	/*
-	 * If we have a last update time, call query_fsmonitor for the set of
-	 * changes since that time, else assume everything is possibly dirty
+	 * If we have a last update token, call query_fsmonitor for the set of
+	 * changes since that token, else assume everything is possibly dirty
 	 * and check it all.
 	 */
 	if (istate->fsmonitor_last_update) {
-		query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION,
-			istate->fsmonitor_last_update, &query_result);
+		if (hook_version == -1 || hook_version == HOOK_INTERFACE_VERSION2) {
+			query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION2,
+				istate->fsmonitor_last_update, &query_result);
+
+			if (query_success) {
+				/* First entry will be the last update token */
+				buf = query_result.buf;
+				bol = 0;
+				for (i = 0; i < query_result.len; i++) {
+					if (buf[i] == '\0') {
+						strbuf_addstr(&last_update_token, buf);
+						bol = i + 1;
+						break;
+					}
+				}	
+			}
```

Comment:
How about adding this to handle failures:

```c
else if (hook_version < 0) {
        hook_version = HOOK_INTERFACE_VERSION1;
        if (!last_update_token.len)
                 		strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
}
```

and also adding `if (hook_version < 0) hook_version = HOOK_INTERFACE_VERSION2;` to the block that handles success? That way, we could avoid sending an initial empty token to a v1 hook.